### ISSUE TITLE
Active-response can't execute ps1 scripts natively

### ISSUE THREAD

**Issue by okynos** [wazuh/wazuh#3002]
Hi team,

We are testing `active-response` module on Windows and we have found an error when we try to execute a `Powershell` `ps1` script when alert of some level is fired, we can see this error on ossec.log when the script is trying to run by Wazuh agent `ossec-agent: ERROR: Unable to create active response process.` it seems to be some problem with `Powershell` execution because `CMD` goes well.

We have done a little workaround to launch custom `ps1` script, we have added `script.cmd` to the Wazuh configuration with the call to such script, example: `powershell.exe active-response/bin/myScript.ps1`, this seems to solve the problem but, maybe will be good to add native support.

We have tested the following things over `ps1` script with direct execution on Wazuh:
* `powershell.exe echo "It doesn't work!" >> C:\file.txt`
* `C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe echo "It doesn't work!" >> C:\file.txt`

**Update**
* `<executable>powershell.exe -file C:\\tmp\\script.ps1</executable>`
* `<executable>C:\\Windows\\SysWOW64\\WindowsPowerShell\\v1.0\\powershell.exe C:\\tmp\\script.ps1</executable>` this prints the message `ossec-agent: INFO: Active response command not present: 'active-response/bin/C:\\Windows\\SysWOW64\\WindowsPowerShell\\v1.0\\powershell.exe C:\\tmp\\script.ps1'. Not using it on this system.` on agent side.

None of them do the job (seems something related with file extension and binary that Wazuh calls).

|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.9.0-3911 | Agent | Packages | Windows Server 2016 |

**Manager `ossec.conf`** (We only have added this part to default `ossec.conf`):
```
  <command>
    <name>test</name>
    <executable>script.ps1</executable>
  </command>

  <active-response>
    <command>test</command>
    <location>local</location>
    <level>3</level>
  </active-response>

```

**Agent `ossec.conf`** (We only have added `active-response`):
```
<ossec_config>

  <client>
    <server>
      <address>192.168.121.100</address>
      <port>1514</port>
      <protocol>tcp</protocol>
    </server>
    <crypto_method>aes</crypto_method>
    <notify_time>10</notify_time>
    <time-reconnect>60</time-reconnect>
    <auto_restart>yes</auto_restart>
  </client>

  <localfile>
    <location>C:\tmp\flood.log</location>
    <log_format>syslog</log_format>
  </localfile>

  <active-response>
    <disabled>no</disabled>
    <ca_store>wpk_root.pem</ca_store>
    <ca_verification>yes</ca_verification>
  </active-response>

  <logging>
    <log_format>plain</log_format>
  </logging>
</ossec_config>
```

Regards.

**Comment by okynos** [comment]
We need to discuss more about this issue.

**Comment by crolopez** [comment]
Hi @okynos,

@SalvaMilenario implemented a version of execd (https://github.com/wazuh/wazuh/pull/3578) that included the powershell alias in the defined command when it is detected to be a powershell script by checking the extension. However, we have decided not to include this feature because it can cause inconsistencies with other types of scripts.

For example, in the case of Python scripts, we would have to check for `.py` extensions, in the case of Perl we would look for `.pl`, `.rb` extensions for Ruby... Also, the scripts do not have to include these typical extensions to be executed.

On the other hand, users may not have the powershell alias configured to be autodetected with `powershell.exe`, or they may have another one that indicates the particular version, such as `powershellv5.exe`. 

For this reason, we believe that only scripts that are native to Windows (written in batch) should be executed without an interpreter. Powershell is a non-native interpreter and does not have to be installed (although it is usually installed in the most recent versions of Windows).

If the user has the powershell alias and wants to use the version of the linker interpreter:

``` XML
  <command>
    <name>test</name>
    <executable>powershell script.ps1</executable>
  </command>
```

If the user, who has or not the alias configured, wants to run a script using powershell 4.0:

``` XML
  <command>
    <name>test</name>
    <executable>C:\Windows\System32\WindowsPowershell\v4.0\powershell.exe script.ps1</executable>
  </command>
```

In short, we think that we should not force to use a powershell version or alias in any case.

Best regards.

**Comment by okynos** [comment]
Hello team,

Thanks for your analysis @crolopez but, we have tested again such situation with your indications without success, it seems to be a problem with how Wazuh manage the command, it can't use another interpreter than CMD.

Regards.

**Comment by crolopez** [comment]
Hi @okynos,

The explanation I gave you is not valid since active-response is ready only to launch executables in `active-response/bin`, and the Powershell script is not considered executable (it need an interpreter). There is no way to specify from the active-response block the interpreter to use from the agent.

The only way to execute a powershell script from an active-response is through a cmd. If you don't do this, the execution of the active response will possibly end up on this line:

https://github.com/wazuh/wazuh/blob/68aaf6d31557106bc8b3cb58b81388644cb27ff5/src/os_execd/exec.c#L176

However, allowing interpreters to be included in the active-response configuration is not currently in the project roadmap, so you will have to use the proposed solution. The active responses will be restructured, simplified and improved in the future, including the possibility of adding interpreters.

Best regards.

**Comment by itpropaul** [comment]
Please address this issue. Needing to call PowerShell scripts from CMD is a HUGE hassle.

I appreciate all of your team's work!

**Comment by dariommr** [comment]
Hello @itpropaul 
Have you tried `ps2exe` PowerShell module? https://www.powershellgallery.com/packages/ps2exe/1.0.4
```
Install-Module ps2exe

## Use the full command name
Invoke-ps2exe .\source.ps1 .\target.exe

## Use the alias
ps2exe .\source.ps1 .\target.exe
```
With this, you will have an `exe` file and you will be able to run it with Active Responses.
It is not a solution but is a powerful workaround.

I hope this could be helpful.

**Comment by itpropaul** [comment]
@dariommr thanks, but that comment was from 3 years ago. Been a minute since I've touched wazuh

**Comment by branchnetconsulting** [comment]
Wazuh seriously needs native PowerShell interpreter support in the exec AR module.  It gets complicated if you want to build a stateful Windows AR, having to bring into consideration the bidirectional signaling between the exec module and the PowerShell script when an intervening cmd script has to be inserted in the middle.  How about just having the exec module notice if an AR command file name ends with a .ps1 extension and honoring that by invoking the script via the PowerShell interpreter natively?

**Comment by ColdAssass1n** [comment]
I know this might sound like a crazy idea, but can you execute shortcuts from active response, particularly .lnk files for windows.

My thought was create a shortcut to powershell.
Goto C:\Windows\SysWOW64\WindowsPowerShell\v1.0\powershell.exe, right click the .Exe and select Send-To Desktop (create shortcut). Take that shortcut (.lnk) and move it into ossec/active-response/bin and rename the .lnk (shortcut) to powershell.lnk Now when you called powershell.exe or cmd file previously in the wazuh active response config, call this:

powershell.lnk -ExecutionPolicy Bypass -file C:\OSSEC-Scripts\someps1file.ps1 -NoProfile -NoNewWindow

Haven't tried it yet but its a thought to an outside-the-box "interpreter"

**Comment by zbalkan** [comment]
> For this reason, we believe that only scripts that are native to Windows (written in batch) should be executed without an interpreter. Powershell is a non-native interpreter and does not have to be installed (although it is usually installed in the most recent versions of Windows).

I am here to do some fact checking. Powershell has been a installed by default since [Windows 7 SP1 and Windows Server 2008 R2](https://web.archive.org/web/20090402204400/http://blogs.msdn.com/powershell/archive/2008/10/28/powershell-will-be-installed-by-default-on-windows-server-08-r2-ws08r2-and-windows-7-w7.aspx) , started being distributed on all Windows devices since around 2008. It was the native shell for more than 10 years when the quoted comment was written. It's almost 14 years now. It is as native for Windows as Bash for Linux, even though there might be other shells on Linux distros. So it is more reliable than other alternatives.

There may be many reasons, architectural decisions, technical issues, etc, to reject Powershell as an AR action. But the case of "not being native" has no factual foundations.

Considering other comments of deploying python as executables are also unrelated to the hardened corporate network clients. Those Windows devices, that we already use in public and private organizations, are mostly configured to run only signed and whitelisted applications via WDAC and AppLocker. The suggestion requires us to either sign the binaries, which is not possible, or add new exceptions to our whitelist for each binary executable we create*. In this case providing a PowerShell script is easier to maintain as it is well known by administrators and security professionals, every Powershell execution is logged and, surprisingly, easily monitored via Wazuh. They can be digitally signed using internal CA, instead of buying a code signing certificate from Microsoft. 

In 2022, it is sad that Powershell cannot be used directly as AR. I was hoping to discuss Powershell 7.x, which is cross platform and can be on any OS, to be included within AR. Yet, here we are. 

As a user of Wazuh, creating content, PRs, issues, asking and questions in Slack, Reddit and mail group, I wholeheartedly hope Wazuh gets better and moves further. And I believe it is possible only by listening to the users and enabling change. 


* It's possible to allow by path and configure `C:\Program Files (x86)\ossec-agent\active-response\`, but it means any malicious file dropped in directory will be trusted. An easy bypass.

**Comment by dariommr** [comment]
Hello All,
I have this way of executing PowerShell scripts from active responses inside batch files without the need of calling a `ps1` file.
We can achieve it by using this inside of the batch file:
```
powershell -executionpolicy ByPass ^
    $alert = Read-Host; ^
    $alert_dict = ConvertFrom-Json $alert; ^
    <MORE LINES>; ^
    $log_line = (Get-Date).ToString('yyyy-MM-dd hh:mm:ss')+\" active-response/bin/remove-file-fim.cmd: AR was executed\"; ^
    Add-Content -Path 'C:\Program Files (x86)\ossec-agent\active-response\active-responses.log' -Value $log_line
```

As you can see, we use the caret `^` to split the lines and the semicolon `;` to specify the end of a command. Also, for the double quotes `"` you need to escape them, but you can use the single quotes without the need of escaping them.

I hope this could be helpful until we can run `ps1` scripts natively.

**Comment by branchnetconsulting** [comment]
So essentially you can run multi-line Powershell from within a .cmd file by starting with this line
`powershell -executionpolicy ByPass ^`
followed by as many lines of Powershell code as desired, where each line of Powershell except the final one is appended with `; ^`

That's a pretty cool workaround, @dariommr !

Until Wazuh properly gets native support for direct AR invocation of ps1 files, this may be just what we need.

After toying with this technique a bit, I see that it scales up even to 150+ lines / 23,000+ characters of Powershell code.

One catch I found was that in certain contexts, you can't use double quotes in the Powershell code.  For example, a line like this does not work
`$A="testing one two three"; ^`
but this does
`$A='testing one two three'; ^`

Next question is if we could generalize this for Windows Powershell stateful active responses, derived from this example:
https://documentation.wazuh.com/current/user-manual/capabilities/active-response/custom-active-response.html#custom-active-response-linux-example

I think the community needs a clear example of a custom stateless and stateful Powershell active response available in the public documentation -- something people could copy and retrofit to their specific purposes as needed.

Kevin

**Comment by branchnetconsulting** [comment]
In case it is useful to anyone else, here is a one-liner to convert a normal ps1 script file into cmd-file embeddable form.  Just make sure to avoid the use of double quotes around strings as that does not work with this embedding technique, at least where string variable assignments are concerned.  Also hash-prefixed comment lines don't appear to work either.

```
$ cat test.ps1
$A=1
$B='hello there'
echo $B
echo $A
whoami

$ echo "powershell -executionpolicy ByPass ^"; head -n -1 test.ps1 | sed 's/$/; ^/;s/^/   /'; tail -n1 test.ps1 | sed 's/^/   /'
powershell -executionpolicy ByPass ^
   $A=1; ^
   $B='hello there'; ^
   echo $B; ^
   echo $A; ^
   whoami
```

**Comment by ColdAssass1n** [comment]
How would I implement this?

**Comment by zbalkan** [comment]
It probably would not work because the json parsing is currently not implemented. This is just an example and I believe it should be handled by Wazuh, so this is a workaround unless it gets into Wazuh source code.

I am on mobile now but it should be something like this ideally. 

```xml
<command>
  <name>run_ps_script</name>
  <executable>ps.py</executable>
  <extra_args>script.ps1</extra_args>
</command>
```

**Comment by bland3man** [comment]
@zbalkan

So would I need the powershell script to be in a certain path on my windows station?

**Comment by branchnetconsulting** [comment]
Wazuh uses Python but I believe Python is only embedded into the wazuh-manager service, and is not embedded into Wazuh Agent, so you would need to have Python installed on Windows endpoints to use this Powershell command runner script, unless of course you compile it into a Windows executable with something like Pyinstaller.  Of course in that case you might as well compile your Powershell script directly into an exe using a similar tool.

**Comment by branchnetconsulting** [comment]
There are security implications to creating a custom active response program that can be used to call a script or run commands that are passed to it as parameters.   If you do this, then anyone with control of your Wazuh manager would be able to run any arbitrary code with full administrative privileges (Local System account) on all Windows agent systems where that custom active response program is present.  This is because the Wazuh API can be used to manually invoke any active response program that exists in the `c:\Program Files (x86)\ossec-agent\active-response\bin\` directory on any Wazuh agent, without an alert triggering it, even if a given program in that directory is not referenced in any way in ossec.conf on that agent or on the manager.
